### PR TITLE
Standardize edge-on axis locations when viewing primary 3d axis planes

### PR DESCRIPTION
**Pull request by scottshambaugh**
## PR Summary
For the special cases of looking at the 3d axes in the 6 primary view planes (defined in the docs in https://github.com/matplotlib/matplotlib/pull/23600), draw the axes such that one is on the bottom edge and the other intersects at the low value (ie the intersection is at the origin for these examples). This blends smoothly when rotating the axes manually, since we are only looking at the epsilon case.

Code to generate these plots is the same as [here](https://github.com/matplotlib/matplotlib/issues/23544#issuecomment-1204292551).

**Before:**
before

**After:**
image

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by QuLogic** [comment]
Would it make sense for the negative planes to swap left/right edges? So that the lines always go through 0.0?

**Comment by scottshambaugh** [comment]
I had thought about that and thought it might be less useful when the axes are shifted away from 0, but could be convinced it’s better. Either is an improvement over the current.

**Comment by scottshambaugh** [comment]
@QuLogic trying it out, I like the layout of that just as much and it probably makes more intuitive sense. I replaced the image in the top post for viewing

**Comment by scottshambaugh** [comment]
Just for fun, here's an "unrolled" view of a 3D axis that might be useful to people
```
import matplotlib.pyplot as plt

# (view, (elev, azim, roll))
views = [( 'XY', (90, -90, 0)), 
         ( 'XZ', (0, -90, 0)),  
         ( 'YZ', (0, 0, 0)), 
         ('-XY', (-90, 90, 0)), 
         ('-XZ', (0, 90, 0)),
         ('-YZ', (0, 180, 0)), 
        ]

layout = [[  '.',  '.', 'XY', '.'],
          ['-XZ','-YZ', 'XZ','YZ'],
          ['-XY',  '.',  '.', '.']]
fig, axd = plt.subplot_mosaic(layout, subplot_kw={'projection': '3d'})
for i in range(len(axd)):
    label = views[i][0]
    axd[label].set_xlabel('x')
    axd[label].set_ylabel('y')
    axd[label].set_zlabel('z')
    axd[label].set_title(label)
    axd[label].set_proj_type('ortho')
    axd[label].view_init(elev=views[i][1][0], azim=views[i][1][1], roll=views[i][1][2])
plt.tight_layout()

```

image

**Comment by QuLogic** [comment]
OK, sorry one more question; can third third Axis also always appear around 0/0?

**Comment by scottshambaugh** [comment]
Unfortunately not, to do so would require having the backing planes in the +-XY and +-YZ views blocking either the plot volume or the axis spines. Here are the plots rotated slightly off-axis to show this. I mean I supposed you could carve out blocking views for specifically these angles, but then it's a discontinuous flash when rotating the plots manually or programmatically.

image

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 1]
```diff
@@ -60,6 +60,26 @@ def test_axes3d_repr():
...
+@mpl3d_image_comparison(['axes3d_primary_views.png'])
+def test_axes3d_primary_views():
+    # (elev, azim, roll)
+    views = [(90, -90, 0),  # XY
+             (0, -90, 0),   # XZ
+             (0, 0, 0),     # YZ
+             (-90, 90, 0),  # -XY
+             (0, 90, 0),    # -XZ
+             (0, 180, 0)]   # -YZ
+    # When viewing primary planes, draw the two other axes on left and bottom
```
Comment (scottshambaugh) [inline lib/mpl_toolkits/tests/test_mplot3d.py:72]:
```suggestion
    # When viewing primary planes, draw the two visible axes so they intersect
    # at their low values
```

**Review by oscargus — APPROVED** [review]

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 2]
```diff
@@ -60,6 +60,27 @@ def test_axes3d_repr():
...
+    views = [(90, -90, 0),  # XY
+             (0, -90, 0),   # XZ
+             (0, 0, 0),     # YZ
+             (-90, 90, 0),  # -XY
+             (0, 90, 0),    # -XZ
+             (0, 180, 0)]   # -YZ
+    # When viewing primary planes, draw the two visible axes so they intersect
+    # at their low values
+    fig, axs = plt.subplots(2, 3, subplot_kw={'projection': '3d'})
+    for i, ax in enumerate([ax for ax_row in axs for ax in ax_row]):
```
Comment (QuLogic) [inline lib/mpl_toolkits/tests/test_mplot3d.py:75]:
```suggestion
    for i, ax in enumerate(axs.flat):
```

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 3]
```diff
@@ -244,11 +244,23 @@ def _get_coord_info(self, renderer):
...
+        # Special handling for edge-on views
+        equals = np.abs(means_z0 - means_z1) <= np.finfo(float).eps
+        if np.sum(equals) == 2:
+            vertical = np.where(np.invert(equals))[0][0]
+            if vertical == 2:  # looking at XY plane
+                highs = np.array([True, True, highs[2]])
+            elif vertical == 1:  # looking at XZ plane
+                highs = np.array([True, highs[1], False])
+            elif vertical == 0:  # looking at YZ plane
+                highs = np.array([highs[0], False, False])
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:263]: Could re-use `highs`?
```suggestion
            vertical = np.where(~equals)[0][0]
            if vertical == 2:  # looking at XY plane
                highs[:2] = True
            elif vertical == 1:  # looking at XZ plane
                highs[0] = True
                highs[2] = False
            elif vertical == 0:  # looking at YZ plane
                highs[1:] = False
```

Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axis3d.py:263]: This is how I originally did it, but I think the verbosity is a lot more clear and reveals the symmetry.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 4]
```diff
@@ -244,11 +244,23 @@ def _get_coord_info(self, renderer):
...
-            mean_z1 = np.mean(bounds_proj[self._PLANES[2 * i + 1], 2])
-            highs[i] = mean_z0 < mean_z1
+            means_z0[i] = np.mean(bounds_proj[self._PLANES[2 * i], 2])
+            means_z1[i] = np.mean(bounds_proj[self._PLANES[2 * i + 1], 2])
+        highs = means_z0 < means_z1
+
+        # Special handling for edge-on views
+        equals = np.abs(means_z0 - means_z1) <= np.finfo(float).eps
+        if np.sum(equals) == 2:
+            vertical = np.where(np.invert(equals))[0][0]
```
Comment (scottshambaugh) [inline lib/mpl_toolkits/mplot3d/axis3d.py:257]:
```suggestion
            vertical = np.where(~equals)[0][0]
```

**Review by QuLogic — APPROVED** [review]

**Pull request merged**